_posts/2020-12-01-configuring-and-suppressing-warnings.md

```diff
@@ -0,0 +1,192 @@
+---
+layout: blog-detail
+post-type: blog
+by: Lukas Rytz
+title: "Configuring and suppressing warnings in Scala"
+isHighlight: true
+---
+
+Scala 2.13.2 and 2.12.13 introduced the `-Wconf` compiler flag to globally configure reporting of warnings, and the `@nowarn` annotation to locally suppress them. Having more control over compiler warnings makes them a lot more valuable:
+
+* In projects where the build log shows a lot of warnings that are mostly ignored, new helpful warnings can easily go undetected. The new functionality can be used to clean up the build log with manageable efforts and potentially enable fatal warnings (`-Werror` in 2.13, `-Xfatal-warnings` in 2.12). This has happened for example in the Scala compiler and standard library projects in the past few months – thank you, [@NthPortal](https://github.com/NthPortal)!
+* Projects that already use fatal warnings get better utilities to work around corner cases where a warning cannot be avoided. This can allow further `-Xlint` checks to be enabled.
+
+In this post we go through the mechanics of configuring warnings and also look at the new `@nowarn` annotation.
+
+## Warnings from the Scala compiler
+
+We start off by recapping some of the most common scenarios where the Scala compiler emits warnings. A first category is suspicious code that is likely the result of a programming mistake:
+
+```scala
+scala> 1 == ""
+         ^
+       warning: comparing values of types Int and String using `==` will always yield false
+
+scala> def f(x: Int) = { x; x }
+                         ^
+       warning: a pure expression does nothing in statement position;
+       multiline expressions might require enclosing parentheses
+
+scala> def f(x: Int): Unit = return x + 1
+                             ^
+       warning: enclosing method f has result type Unit: return value of type Int discarded
+```
+
+Warnings about non-exhaustive pattern matches and uncheckable type arguments are also issued by default:
+
+```scala
+scala> def get[T](o: Option[T]): T = o match { case Some(t) => t }
+                                     ^
+       warning: match may not be exhaustive.
+       It would fail on the following input: None
+
+scala> def f(l: List[Any]) = l match { case l: List[Int] => l.sum; case _ => l.length }
+                                               ^
+       warning: non-variable type argument Int in type pattern List[Int]
+       is unchecked since it is eliminated by erasure
+```
+
+Annotating the scrutinee with `@unchecked` disables exhaustivity checking: `(o: @unchecked) match ...`. Similarly, unchecked warnings are not issued for annotated type arguments: `case l: List[Int @unchecked] => ...`.
```

Comment:
It wouldn't hurt to mention that exhaustivity checking is much improved in 2.13.4